### PR TITLE
Fix page gap inconsistencies between doc open & config change

### PR DESCRIPTION
**Pull request by NiLuJe**
Specifically, doc open runs the value through `scaleBySize`, as expected, but the event handler did not.

Switch the setting to a buttonprogress to allow using more_options in a less clunky fashion.

Also show the actual value in pixels in the long-press recap & notification.
This involved an extension of some helpers in optionsutil, and made me realize that I find showing these flex pixels as actual device pixels instead of mm/in much much much much more relevant, so, err, RFC to actually do something more involved on that front (expand the mm/in toggle to a mm/in/px tristate? Always show px in addition to mm/in?)

pagegap-buttonprogress

(Rebase me)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12507)

### CODE REVIEW

**Comment by NiLuJe** [comment]
> I find showing these flew pixels as actual device pixels instead of mm/in much much much much more relevant, so, err, RFC to actually do something more involved on that front (expand the mm/in toggle to a mm/in/px tristate? Always show px in addition to mm/in?)

e.g.,

Screenshot_20240912_003513

vs.

Screenshot_20240912_003528


(In both case, the input is a "koreader flex" pixel, unless I hallucinated something).

**Comment by NiLuJe** [comment]
(c.f., #9205 for more context about the above)

**Comment by NiLuJe** [comment]
(The above screenshots were missing a `default_pos`, so just imagine the default square having a black border as expected ;p).

File: frontend/ui/data/koptoptions.lua [hunk 1]
```diff
@@ -349,16 +349,25 @@ left to right or reverse, top to bottom or reverse.]]),
...
                 end,
                 name_text_hold_callback = optionsutil.showValues,
+                name_text_true_values = true,
+                name_text_unit = "px", -- values go through scaleBySize in the event handler
                 help_text = _([[In continuous view mode, sets the thickness of the separator between document pages.]]),
+                more_options = true,
+                more_options_param = {
+                    value_step = 1, value_hold_step = 10,
+                    value_min = 0, value_max = 256,
+                    precision = "%.1f",
```
Comment (hius07) [inline frontend/ui/data/koptoptions.lua:369]: Decimal point?

**Comment by Frenzie** [comment]
Real pixels sounds useful, yes. The trouble with Wayland and macOS is that you don't seem to get any, guaranteeing eternal blurriness.

File: frontend/ui/data/koptoptions.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/ui/data/koptoptions.lua:369]: Yeah, necessary if you want to be able to icnrese in precise 'absolute) 1px steps on high-dpi devices and/or on the bigger end of the scale.

e.g., on a Clara 2E, 255 is something like 456px, and 256 458px. If you really really wanted 457px, being able to enter 455.5 helps doing just that ;).

Since it should be pretty niche, I left the step in integers, as that's what most people should be using ;).

TL;DR: `scaleBySize` handles floats just fine, so it can be useful ;).

**Comment by NiLuJe** [comment]
> Real pixels sounds useful, yes. The trouble with Wayland and macOS is that you don't seem to get any, guaranteeing eternal blurriness.

My current line of thought would be to reword the `Metric length` entry in the `Units` section of the menu to read something like "Dimensions" instead, and to show a RadioButtonWidget with three choices: Metric/Imperial/Pixel

**Comment by NiLuJe** [comment]
> My current line of thought would be to reword the `Metric length` entry in the `Units` section of the menu to read something like "Dimensions" instead, and to show a RadioButtonWidget with three choices: Metric/Imperial/Pixel

As opposed to showing the px value *in addition* to the in/mm value.

**Comment by NiLuJe** [comment]
Done:

Screenshot_20240912_210237
Screenshot_20240912_210245
Screenshot_20240912_210256

**Comment by poire-z** [comment]
> > My current line of thought would be to reword the `Metric length` entry in the `Units` section of the menu to read something like "Dimensions" instead, and to show a RadioButtonWidget with three choices: Metric/Imperial/Pixel
> 
> As opposed to showing the px value _in addition_ to the in/mm value.

Somehow, as much as I can get one only _speaks_ metric `cm` or english `inches` and rarely both, each population can also understand `px`. I guess I would be happy keep seeing both `mm` and `px`, and a bit sad having to choose only one of them.

**Comment by NiLuJe** [comment]
It *should* be fairly easy to make that possible (and optional, even) with that refactor, lemme look into it ;).

**Comment by NiLuJe** [comment]
Yup, neat and easy ;).

Screenshot_20240912_211837
Screenshot_20240912_211846

**Comment by mergen3107** [comment]
To add to consistency, there should be a unit for the first value as well. 

10 what? 10 units? 10 KOReader-scale units? 10 K-units? :D

**Comment by Frenzie** [comment]
> 10 KOReader-scale units

Correct, essentially that which would be 1 pixel on a 160 DPI screen, but typically it'll be something like twice the value on most newer devices.

**Comment by poire-z** [comment]
I think we don't need to mention any unit for our abstract number font size.
It's like the volume on our TVs: we see a number on the screen, we soon get what that number means for us.

**Comment by NiLuJe** [comment]
In case another sleep-deprived brain gets confused by the appended px values for font sizes in pt: remember that the DPI is not accurate on the emu, so the pt result will be bogus (as that depends on the ppi being accurate).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 2 of 5 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, 3 of 3 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12507)* status: all files reviewed, 1 unresolved discussion (waiting on @hius07)

File: frontend/apps/reader/modules/readertypeset.lua [hunk 2]
```diff
@@ -506,13 +506,19 @@ function ReaderTypeset:onSetPageMargins(margins, when_applied_callback)
...
+
+  footer: %5 px
 
 Tap to dismiss.]]),
-            margins[1], left, margins[3], right, margins[2], top, margins[4], bottom),
+            optionsutil.formatFlexSize(margins[1]),
+            optionsutil.formatFlexSize(margins[2]),
+            optionsutil.formatFlexSize(margins[3]),
+            optionsutil.formatFlexSize(margins[4]),
+            self.view.footer.reclaim_height and 0 or self.view.footer:getHeight()),
```
Comment (poire-z) [inline frontend/apps/reader/modules/readertypeset.lua:521]: No flex size for the footer height ?

File: frontend/ui/data/optionsutil.lua [hunk 3]
```diff
@@ -17,52 +17,85 @@ function optionsutil.enableIfEquals(configurable, option, value)
...
         shown_unit = C_("Font size", "pt")
     elseif unit == "mm" then
         shown_unit = C_("Length", "mm")
     elseif unit == "in" then
         shown_unit = C_("Length", "in")
-    else
-        shown_unit = unit -- for future units
+    elseif unit == "px" then
+        shown_unit = C_("Pixels", "px")
+        -- We don't so subpixel positioning ;)
```
Comment (poire-z) [inline frontend/ui/data/optionsutil.lua:67]: We so don't _do_ subpixel positioning ;)

File: frontend/ui/data/creoptions.lua [hunk 4]
```diff
@@ -221,9 +221,8 @@ In the top menu → Settings → Status bar, you can choose whether the bottom m
                     G_defaults:readSetting("DCREREADER_CONFIG_T_MARGIN_SIZES_XX_HUGE"),
                 },
                 hide_on_apply = true,
-                name_text_hold_callback = function(configurable, opt, prefix)
-                    optionsutil.showValues(configurable, opt, prefix, nil, "mm")
-                end,
+                name_text_hold_callback = optionsutil.showValues,
+                name_text_unit = true,
```
Comment (poire-z) [inline frontend/ui/data/creoptions.lua:225]: I don't think I've read a comment anywhere saying what `true` means.

**Review by poire-z — APPROVED** [review]
Trusting you on the maths.

File: frontend/apps/reader/modules/readertypeset.lua (hunk 2 above)
Comment (NiLuJe) [reply frontend/apps/reader/modules/readertypeset.lua:521]: Nope, it's already in pixels ;)).

File: frontend/ui/data/creoptions.lua (hunk 4 above)
Comment (NiLuJe) [reply frontend/ui/data/creoptions.lua:225]: Will do. (It basically piggybacks on the `unit and unit ~= "pt" -> use user setting` check in optionsutil to mean "it's a dimension that isn't in pt, follow user settings")

File: frontend/apps/reader/modules/readertypeset.lua (hunk 2 above)
Comment (poire-z) [reply frontend/apps/reader/modules/readertypeset.lua:521]: Like the margins, but we show them in cm/in - so why not the footer height?

Comment (NiLuJe) [reply frontend/apps/reader/modules/readertypeset.lua:521]: The variables being used for the margins here is pre-scaling, so they're still flex, not absolute pixels like the footer:

https://github.com/NiLuJe/koreader/blob/bfecf88f38271131d3fde701f68be0d169ff3011/frontend/apps/reader/modules/readertypeset.lua#L490-L498

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12507)* status: all files reviewed, 3 unresolved discussions (waiting on @poire-z)

___
*[`frontend/ui/data/creoptions.lua` line 225 at r6](https://reviewable.io/reviews/koreader/koreader/12507#gh-1759356052:-O6lsHlo-Z5mqwTYNf5x:b-896fix) ([raw file](https://github.com/koreader/koreader/blob/bfecf88f38271131d3fde701f68be0d169ff3011/frontend/ui/data/creoptions.lua#L225)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

Will do. (It basically piggybacks on the `unit and unit ~= "pt" -> use user setting` check in optionsutil to mean "it's a dimension that isn't in pt, follow user settings")
</blockquote></details>

Done.
___
*[`frontend/ui/data/optionsutil.lua` line 67 at r6](https://reviewable.io/reviews/koreader/koreader/12507#gh-1759350872:-O6lsKQHCvkOxmcuTwgU:b-896fix) ([raw file](https://github.com/koreader/koreader/blob/bfecf88f38271131d3fde701f68be0d169ff3011/frontend/ui/data/optionsutil.lua#L67)):*
<details><summary><i>Previously, poire-z wrote…</i></summary><blockquote>

We so don't _do_ subpixel positioning ;)
</blockquote></details>

Done.

**Comment by yparitcher** [comment]
@NiLuJe 
Can you double check the the toggle / button change does not break existing dispatcher items?

**Comment by NiLuJe** [comment]
> Can you double check the the toggle / button change does not break existing dispatcher items?

The args haven't changed, so everything should still match.

Before:

Screenshot_20240915_202702

After:

Screenshot_20240915_202729
Screenshot_20240915_202754
